### ISSUE TITLE
min and maxRenderIndex broken when using sampleRate/samplePeriod in SignalPlot?

### ISSUE THREAD

**Issue by LB767** [swharden/ScottPlot#621]
When using a sampleRate different than 1 in SignalPlot & SignalPlotConst the min and maxRenderIndex properties don't seem to behave properly anymore.
I tried to play with the sampleRate value to see if I could see a pattern that made sense (like min and max being double for a sampleRate of 2 or something) but I couldn't see anything...

Just try
```cs
plt.PlotSignalConst(ys, sampleRate: 2, minRenderIndex: 100, maxRenderIndex: 200);
```
and at least for me this doesn't work
(Using the latest Scottplot 4.0.42 with WPF)

**Comment by swharden** [comment]
Hi @LB767, 

I believe `minRenderIndex` and `maxRenderIndex` define the _index_ in the `ys` array to display. It is not sensitive to sample rate. I hope this helps!

**Comment by LB767** [comment]
Yes, that is exactly what I expected, but it doesn't work like that. As I mentionned the displayed data changes based on the samplerate value, and with weird behavior, sometimes it doesn't display any data all, as if the min and max ended up on the same index

**Comment by swharden** [comment]
Hey @LB767, I tested this by writing code that plots 1k points with a sample rate of 10 (so 1,000 points spans 0-100 on the horizontal axis). The output renders as I would expect.

Can you provide a code sample that produces output different than you'd expect? Perhaps I'm misunderstanding your concern.

```cs
Random rand = new Random(0);
double[] ys = ScottPlot.DataGen.RandomWalk(rand, 1000);

var plt = new ScottPlot.Plot();
var sig = plt.PlotSignal(ys, sampleRate: 10);

plt.Title("no limit");
plt.SaveFig("1.png");

sig.minRenderIndex = 450;
sig.maxRenderIndex = ys.Length - 1;
plt.Title("lower data removed");
plt.SaveFig("2.png");

sig.minRenderIndex = 0;
sig.maxRenderIndex = 550;
plt.Title("upper data removed");
plt.SaveFig("3.png");

sig.minRenderIndex = 450;
sig.maxRenderIndex = 550;
plt.Title("lower and upper data removed");
plt.SaveFig("4.png");
```

![image](https://user-images.githubusercontent.com/4165489/99888479-29aee900-2c1b-11eb-8277-f793dfbccccb.png)

![image](https://user-images.githubusercontent.com/4165489/99888477-26b3f880-2c1b-11eb-9cbc-56e0d9b1b69d.png)

![image](https://user-images.githubusercontent.com/4165489/99888476-23207180-2c1b-11eb-8d9c-e547968f9c23.png)

![image](https://user-images.githubusercontent.com/4165489/99888475-20258100-2c1b-11eb-8637-83bd161e9a2d.png)

Hope this helps!
Scott

**Comment by LB767** [comment]
Hey @swharden , thank you for taking the time to try this out on your side. Looks like there's more head-scratching involved than I assumed....
I copy-pasted your code, and sure enough it worked exactly as it should, so I tried stripping it down and apparently this works
```cs
Random rand = new Random(0);
double[] ys = ScottPlot.DataGen.RandomWalk(rand, 1000);

var plt = new ScottPlot.Plot();
var sig = plt.PlotSignal(ys, sampleRate: 10);
                          
plt.SaveFig("1.png");      
  
sig.minRenderIndex = 450;
sig.maxRenderIndex = 550;
plt.Title("lower and upper data removed");
plt.SaveFig("4.png");
```

But this **doesn't**, and produces a completely empty plot:
```cs
Random rand = new Random(0);
double[] ys = ScottPlot.DataGen.RandomWalk(rand, 1000);

var plt = new ScottPlot.Plot();
var sig = plt.PlotSignal(ys, sampleRate: 10);
                          
sig.minRenderIndex = 450;
sig.maxRenderIndex = 550;
plt.Title("lower and upper data removed");
plt.SaveFig("4.png");
```

I frankly have no idea what's going on here, it almost seems like it's more to do with the layout of the plot where something might get set when calling `SaveFig()` which goes into a weird state if it's called with `minRenderIndex `& `maxRenderIndex ` set. No clue... I'm gonna try and play with it more in the meantime, see if I can find any lead...

Thanks again for your time

**Comment by swharden** [comment]
I think this a bug related to how automatic axis limits are detected in signal plots with min/max render index specified. I added some code to the end of your example and it seems that the X limits are broken (with the upper limit being a smaller number than the lower limit)

```cs
Console.WriteLine($"Axis Limits: {new ScottPlot.Config.AxisLimits2D(plt.Axis())}");
```

```
Axis Limits: x1=471.944, x2=33.056, y1=13.876, y2=23.248
```

A quick fix is to call AxisAuto() before setting the min/max render index. I recognize this isn't a great solution though, because middle-clicking in a user control will call AxisAuto() later and the problem will happen again.

```cs
Random rand = new Random(0);
double[] ys = ScottPlot.DataGen.RandomWalk(rand, 1000);

var plt = new ScottPlot.Plot(400, 300);
var sig = plt.PlotSignal(ys, sampleRate: 10);

plt.AxisAuto(); // set good limits before hiding data

sig.minRenderIndex = 450;
sig.maxRenderIndex = 550;
plt.Title("lower and upper data removed");

plt.SaveFig("4.png");
```

![image](https://user-images.githubusercontent.com/4165489/99913820-c29b3e00-2cc7-11eb-820c-ea4aa493922b.png)

I think the problem relates to this line.

https://github.com/swharden/ScottPlot/blob/4bf514cfe65e19d644db25f961a81f30e44caa2d/src/ScottPlot/plottables/PlottableSignalBase.cs#L198

I see setting the sample rate updates the sample period, but updating the sample period does not update the sample rate!

https://github.com/swharden/ScottPlot/blob/4bf514cfe65e19d644db25f961a81f30e44caa2d/src/ScottPlot/plottables/PlottableSignalBase.cs#L43-L67

I'll continue to explore this...

**Comment by swharden** [comment]
I just published 4.0.44 on NuGet which should have this bug corrected!

AxisAuto() still doesn't ignore invisible plots, but the code example above should get the job done 👍
